Json append fix

pranavsinghps1 opened the pull request:
With respect to ```vissl/utils/io.py``` file, on using ```append_to_json=True``` in the ```save_file``` function (in its current state) appends another JSON object to an already existing one rendering the file unreadable.This also means that the append function is not behaving as it is meant to.

To reproduce the above, run the https://colab.research.google.com/github/facebookresearch/vissl/blob/stable/tutorials/Understanding_VISSL_Training_and_YAML_Config.ipynb in its current state, and check the ```dataset_catalog.json``` after running the cell https://colab.research.google.com/github/facebookresearch/vissl/blob/stable/tutorials/Understanding_VISSL_Training_and_YAML_Config.ipynb#scrollTo=M8Q6LCqaWjl1&line=10&uniqifier=1 , the very next cell raises an error.
Also mentioned in #373 , the remedy to this is to delete the ```dataset_catalog.json``` file as mentioned in https://github.com/facebookresearch/vissl/issues/373#issuecomment-889586387

With this PR, I have made changes to the append functionality so that it writes to the already existing JSON object instead of creating two (basically without destabilizing the JSON format) and thereby actually achieving the appending functionality. Demo of this: https://colab.research.google.com/drive/1DMNshw4NJGi2WmFq6x04YP-kgSJWV6QU?usp=sharing

iseessel commented:
```
With respect to vissl/utils/io.py file, on using append_to_json=True in the save_file function (in its current state) appends another JSON object to an already existing one rendering the file unreadable.This also means that the append function is not behaving as it is meant to.
```

To me, this seems like the right functionality for `append_to_json`. (We use this in populating our `metrics.json` file, which we want to append a new json object to the file). 

I think the proper solution to https://github.com/facebookresearch/vissl/issues/373#issuecomment-889586387 would be to change the tutorial to add append_json=False to all the save_file calls in the tutorials. What do you think?

pranavsinghps1 commented:
> ```
> With respect to vissl/utils/io.py file, on using append_to_json=True in the save_file function (in its current state) appends another JSON object to an already existing one rendering the file unreadable.This also means that the append function is not behaving as it is meant to.
> ```
> 
> To me, this seems like the right functionality for `append_to_json`. (We use this in populating our `metrics.json` file, which we want to append a new json object to the file).
> 
> I think the proper solution to [#373 (comment)](https://github.com/facebookresearch/vissl/issues/373#issuecomment-889586387) would be to change the tutorial to add append_json=False to all the save_file calls in the tutorials. What do you think?

Indeed changing ```append_to_json``` to False would solve the issue, but I guess we should also add comments on how it could be used for multiple entries as well with ```append_to_json = True```, making it easier for beginners but also giving pointers for advanced users.

iseessel commented:
Sure! The format that appending json is called: newline delimited json! feel free to add a comment.

pranavsinghps1 commented:
> Sure -- yeah I see the confusion.
> 
> The format that appending json is called: newline delimited json! feel free to add a comment.

- Fixed the tutorial (tutorials/Understanding_VISSL_Training_and_YAML_Config.ipynb), the save_file function now rewrites instead of appending. (Installation part still needs to be addressed).
- Added comments in Tutorials.

iseessel commented on vissl/utils/io.py:
```diff
@@ -69,9 +69,13 @@ def save_file(data, filename, append_to_json=True):
             np.save(fopen, data)
     elif file_ext == ".json":
         if append_to_json:
```
This needs to be reverted as discussed in comments.

iseessel commented on tutorials/Understanding_VISSL_Training_and_YAML_Config.ipynb:
```diff
@@ -266,7 +266,11 @@
-        "**NOTE**: VISSL uses the specific `dataset_catalog.json` under the path `configs/config/dataset_catalog.json`."
+        "**NOTE**: VISSL uses the specific `dataset_catalog.json` under the path `configs/config/dataset_catalog.json`.\n",
+        "\n",
+        "**NOTE**: If you want to preserve the older entries and create a newline delimited json for `dataset_catalog.json`, use `append_to_json=True` in the `save_file` function. \n",
```
Sorry, I don't think I was super clear in my comments. new_line_delimited json won't work for the dataset_catalog -- only json, we have to use append_to_json=False. I personally don't think we need to add a comment here, as this is really just an implementation detail. 

What I meant by new_line_delimited json was when we save the metrics.json file. See: https://github.com/facebookresearch/vissl/blob/master/vissl/hooks/log_hooks.py#L499
which yields the following format:

```
{"iteration": 312, "phase_idx": 0, "train_phase_idx": 0}
{"iteration": 624, "phase_idx": 1, "train_phase_idx": 1}
{"iteration": 936, "phase_idx": 2, "train_phase_idx": 2}
{"iteration": 1248, "phase_idx": 3, "train_phase_idx": 3}
{"iteration": 1560, "phase_idx": 4, "train_phase_idx": 4}
{"iteration": 1872, "phase_idx": 5, "train_phase_idx": 5}
```

pranavsinghps1 replied:
Addressed both the comments in e6449ba
